### PR TITLE
symcheck files added to monitor if syms are being updated in subscrib…

### PR DESCRIPTION
**Pull request by karanpatel12**
…er - Requires Hardcoded Torq working directory to be changed to the KDBCODE environment variable

### CODE REVIEW

File: code/common/symcheck.q [hunk 1]
```diff
@@ -0,0 +1,13 @@
+a:.Q.opt .z.x
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: should be a TorQ process, therefore just use `.proc.params` which already handles this

File: code/common/symcheck.q [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+h:hopen hsym `$raze string[.z.h],":", (a `port) ,":admin:admin" / open handle h to the RDB  port
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: what if port isn't passed into process?
shouldn't login as admin
string construction is messy
should be using discovery process in TorQ

File: code/common/symcheck.q [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+t:((.z.p-.z.p)+`second$("F"$a `time))[0] /string of time t since the sym was updated
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: Awful lot of brackets, unnecessary whitespace in ``a `time``

File: code/common/symcheck.q [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+
+b:h".z.p -first exec time from select [-1] time from trade"
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: whitespace

File: code/common/symcheck.q [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+if[b > t; 0N! "RDB trade table is no longer updating"; exit 0] / check is rdb table is updating by comparing the time of the last trade to t
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: whitespace, more than one action so should probably be multiline from Thomas' style guidelines

File: code/common/symcheck.q [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+tab:h raze "select sym,lastseen:time,delta:" , (string t) , "<.z.p-time from (select by sym from trade)" / query for the last time for each sym, then check if it is greater than t
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: constructing a query string is messy, use a lambda

File: code/common/symcheck.q [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+if[0<count where 0=tab.delta; show select sym, lastseen from tab where 1=tab.delta] / Only print table to output if the trades table is non-empty
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: don't need `tab.delta` at the end of the line where it's in a select

File: code/common/symcheckwrapper.sh [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+#Must be run in a cronjob
+# Checking the required arguments. Use default arguments if all 3 arguemnts are not given
+
+# QLIC=/opt/kdb/QLIC
+# QHOME=/opt/kdb/3.5/2017.11.30
+echo $HOMEDI
+if [ "$#" -ne 3 ]; then
```
Comment (jonathonmcmurray) [inline code/common/symcheckwrapper.sh]: What if I pass in 2 args and want default for third? Or non-positional arguments?

File: code/common/symcheckwrapper.sh [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+	emailid="karan.patel@aquaq.co.uk" # default emailid for notifications CHANGE THIS
+	t=600	# seconds since the sym has not updated
+	p=9802	# port number of subscriber to check # CHANGE THIS
+else
+	emailid=$3
+	t=$2
+	p=$1
+fi
+# Run q sript and pipe output table into mail. -E flag does not send mail is message body is empty
+/usr/bin/rlwrap /opt/kdb/3.5/2017.11.30/l64/q /home/kpatel/proddev1/TorQ/code/common/symcheck.q -time $t -port $p | mail -Es "Warning: sym is missing" $emailid #NEED TO CHANGE filename to take Torq working directory from $KDBCODE
```
Comment (jonathonmcmurray) [inline code/common/symcheckwrapper.sh]: I'd much rather you sent the email from Q using TorQ's built in functions

Comment (jonathonmcmurray) [reply code/common/symcheckwrapper.sh]: Also, surely this will send an email everytime, even if there aren't any missing syms (would just send an empty email)

**Comment by jonathonmcmurray** [comment]
Still not running as a TorQ process and using TorQ functionality...time's running out...

File: code/common/symcheck.q [hunk 10]
```diff
@@ -1,13 +1,16 @@
-a:.Q.opt .z.x
+def:.Q.def[`stackID`user`pass!(1902;`admin;`admin)].Q.opt[.z.x];
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: Should not be logging as admin (I feel like I've said this about 60 times by now?)

File: code/common/symcheck.q [hunk 11]
```diff
@@ -1,13 +1,16 @@
...
 
-h:hopen hsym `$raze string[.z.h],":", (a `port) ,":admin:admin" / open handle h to the RDB  port
+path:`$"::",string[def[`stackID]],":",string[def[`user]],":",string[def[`pass]];
 
-t:((.z.p-.z.p)+`second$("F"$a `time))[0] /string of time t since the sym was updated
+opHandle:{@[hopen;path;{-2"ERROR: ",x}]};
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: why make this a function and then call it immediately? just do it without any `{}` round it

File: code/common/symcheckwrapper.sh [hunk 12]
```diff
@@ -1,19 +1,40 @@
 #!/bin/bash
-#Must be run in a cronjob
+#Can be run in a cronjob to check repeatedly on timer
+#  EDIT LINE 41 to point to relevant symcheck.q file
```
Comment (jonathonmcmurray) [inline code/common/symcheckwrapper.sh]: nope nope nope

File: code/common/symcheck.q [hunk 13]
```diff
@@ -1,13 +1,16 @@
...
+h:opHandle[];
 
+t:(.z.p-.z.p)+`second$"f"$def[`time] /t is the given input time interval to check the sym was last updated
 
-b:h".z.p -first exec time from select [-1] time from trade"
-if[b > t; 0N! "RDB trade table is no longer updating"; exit 0] / check is rdb table is updating by comparing the time of the last trade to t
+if[t<.z.p-h"last exec time from trade";0N!"RDB trade table is no longer updating"; exit 0] / check if rdb table is updating by comparing the time of the last trade to t
+h(set;`t;value`t); /define t to exist on sever side
+
+tab:h"{[x]select sym,lastseen:time,bSymUpdate:x<.z.p-time from select by sym from trade}t" / query for the last time for each sym, then check if it is greater than t
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: this is not what I meant with constructing a lambda...

something like

`h({...};t)`

will pass over the lambda & the arg

File: code/common/symcheck.q [hunk 14]
```diff
@@ -1,13 +1,16 @@
...
-t:((.z.p-.z.p)+`second$("F"$a `time))[0] /string of time t since the sym was updated
+opHandle:{@[hopen;path;{-2"ERROR: ",x}]};
+h:opHandle[];
 
+t:(.z.p-.z.p)+`second$"f"$def[`time] /t is the given input time interval to check the sym was last updated
 
-b:h".z.p -first exec time from select [-1] time from trade"
-if[b > t; 0N! "RDB trade table is no longer updating"; exit 0] / check is rdb table is updating by comparing the time of the last trade to t
+if[t<.z.p-h"last exec time from trade";0N!"RDB trade table is no longer updating"; exit 0] / check if rdb table is updating by comparing the time of the last trade to t
+h(set;`t;value`t); /define t to exist on sever side
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: you shouldn't be setting a variable on the server for this

File: code/common/symcheck.q [hunk 15]
```diff
@@ -1,13 +1,16 @@
...
 
-t:((.z.p-.z.p)+`second$("F"$a `time))[0] /string of time t since the sym was updated
+opHandle:{@[hopen;path;{-2"ERROR: ",x}]};
+h:opHandle[];
 
+t:(.z.p-.z.p)+`second$"f"$def[`time] /t is the given input time interval to check the sym was last updated
 
-b:h".z.p -first exec time from select [-1] time from trade"
-if[b > t; 0N! "RDB trade table is no longer updating"; exit 0] / check is rdb table is updating by comparing the time of the last trade to t
+if[t<.z.p-h"last exec time from trade";0N!"RDB trade table is no longer updating"; exit 0] / check if rdb table is updating by comparing the time of the last trade to t
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: `0N!` isn't how you should do output, this should be done using `-2` for stderr (or `-1` for stdout, but looks like this is an error)
`exit 0` means the exit code will be 0 i.e. no error. by convention, any time you're exiting a process on an error, you should always return a non-zero exit code (makes it a lot easier to wrap it with bash commands as well, you can do different things based on if it succeeded or failed)

File: code/common/symcheck.q [hunk 16]
```diff
@@ -1,13 +1,16 @@
...
+h(set;`t;value`t); /define t to exist on sever side
+
+tab:h"{[x]select sym,lastseen:time,bSymUpdate:x<.z.p-time from select by sym from trade}t" / query for the last time for each sym, then check if it is greater than t
+
+if[0<count where 1=tab.bSymUpdate; show select sym,lastseen from tab where 1=tab.bSymUpdate] / Only print table to output if the trades table is non-empty
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: do not need `tab.bSymUpdate`, just `bSymUpdate` (as per previous comment)

File: code/common/symcheck.q [hunk 17]
```diff
@@ -1,13 +1,16 @@
...
+def:.Q.def[`stackID`user`pass!(1902;`admin;`admin)].Q.opt[.z.x];
 
-h:hopen hsym `$raze string[.z.h],":", (a `port) ,":admin:admin" / open handle h to the RDB  port
+path:`$"::",string[def[`stackID]],":",string[def[`user]],":",string[def[`pass]];
 
-t:((.z.p-.z.p)+`second$("F"$a `time))[0] /string of time t since the sym was updated
+opHandle:{@[hopen;path;{-2"ERROR: ",x}]};
+h:opHandle[];
 
+t:(.z.p-.z.p)+`second$"f"$def[`time] /t is the given input time interval to check the sym was last updated
```
Comment (jonathonmcmurray) [inline code/common/symcheck.q]: don't need square brackets round `` `test``

Comment (jonathonmcmurray) [reply code/common/symcheck.q]: also `(.z.p-.z.p)` - there is almost certainly a better way to do whatever you're trying to do here (not completely clear what's happening

File: code/common/symcheck.q [hunk 18]
```diff
@@ -0,0 +1,16 @@
+def:.Q.def[`stackID`user`pass!(1902;`admin;`admin)].Q.opt[.z.x];
```
Comment (ThomasSmyth) [inline code/common/symcheck.q]: Trim squares.

File: code/common/symcheck.q [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+
+path:`$"::",string[def[`stackID]],":",string[def[`user]],":",string[def[`pass]];
```
Comment (ThomasSmyth) [inline code/common/symcheck.q]: Square bracket party on this line, but it's closing time. Time to clear them out.

File: code/common/symcheck.q [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+
+opHandle:{@[hopen;path;{-2"ERROR: ",x}]};
+h:opHandle[];
+
+t:(.z.p-.z.p)+`second$"f"$def[`time] /t is the given input time interval to check the sym was last updated
```
Comment (ThomasSmyth) [inline code/common/symcheck.q]: `(.z.p-.z.p)` will return 0, is this to force a timestamp?. Why is time not handled by `.Q.def` above? Would save on this line. + No need for squares + need for ending semi colon.

File: code/common/symcheck.q [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+
+if[t<.z.p-h"last exec time from trade";0N!"RDB trade table is no longer updating"; exit 0] / check if rdb table is updating by comparing the time of the last trade to t
```
Comment (ThomasSmyth) [inline code/common/symcheck.q]: This `if` statement has multiple actions, better to make it multiline to highlight this.

File: code/common/symcheck.q [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+h(set;`t;value`t); /define t to exist on sever side
```
Comment (ThomasSmyth) [inline code/common/symcheck.q]: Why does `t` need set server side, can it not be passed as and when required?

File: code/common/symcheckwrapper.sh [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+do
+  case "${option}"
+  in
+  t) t=${OPTARG}
+     ;;
+  p) p=${OPTARG}
+     ;;
+  e) emailid=${OPTARG}
+     ;;
+ \?)
```
Comment (ThomasSmyth) [inline code/common/symcheckwrapper.sh]: Other cases have inline code, seems inconsistent to change it up for this case.

File: code/common/symcheckwrapper.sh [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+  t=600 # default seconds since the sym has not updated
+fi
+
+if [ -z $p ]
+then
+  p=9802 # default port number of subscriber to check # CHANGE THIS
+fi
+
+# Run q sript and pipe output table into mail. -E flag does not send mail is message body is empty
+/usr/bin/rlwrap /opt/kdb/3.5/2017.11.30/l64/q /home/kpatel/proddev1/TorQ/code/common/symcheck.q -time $t -port $p | mail -Es "Warning: sym is missing" $emailid #NEED TO CHANGE filename to take Torq working directory from $KDBCODE
```
Comment (ThomasSmyth) [inline code/common/symcheckwrapper.sh]: Hardcoded paths should be avoided + rlwrap is only required for interactive sessions.
